It's YAML all the way down!

gilesknap opened the pull request:
This PR adds an suggested approach to using Ansible as the way to script support module and IOC builds inside of epics-containers.

This file https://github.com/epics-containers/ibek-support/blob/ansible2/_ansible/roles/support/vars/main.yml contains the defaults that you override for each support module.

This PR is for ioc-adsimdetector using the feature in full. https://github.com/epics-containers/ioc-adsimdetector/pull/64

Comments welcome!

marcelldls commented:
@gilesknap Very neat! What are some of the benefits of using this approach? I believe some core benefits of Ansible are working with multiple hosts and idempotency, but I don't believe that is realized when working with containers?

If the goal is to support bare metal deployments, would you then be porting all the `RUN` commands from the `ioc-adsimdetector/Dockerfile` to Ansible too? It seemed to me like the Epics training Ansible had to jump through a significant amount of hoops to support different Distros

gilesknap commented:
> @gilesknap Very neat! What are some of the benefits of using this approach? I believe some core benefits of Ansible are working with multiple hosts and idempotency, but I don't believe that is realized when working with containers?
> 
> If the goal is to support bare metal deployments, would you then be porting all the `RUN` commands from the `ioc-adsimdetector/Dockerfile` to Ansible too? It seemed to me like the Epics training Ansible had to jump through a significant amount of hoops to support different Distros

My reasons for using Ansible are:-
- get rid of bespoke Python code that needs supporting and explaining
- use idempotency when making use of this inside of developer containers
- get more control of the overall build process (the existing approach has lots of bits of script that rely on each other e.g. every install.sh is required to call the _globals/install.sh and if you don't subtle issues will appear)
- provide a 'rebuild all support' inside the devcontainer (this is TODO)

I don't plan to remove the individual RUN commands because they make the container cache behave nicely. For those that don't care about that (but you really should) they could make one RUN that calls the 'rebuild all support' I mentioned.

I was not really intending to reuse this outside of containers but maybe that idea has some merit.

coretl commented on ADCore/ADCore.yml:
```diff
@@ -0,0 +1,85 @@
+module: ADCore
```
Any possiblity to make a schema for these files? Also, could we put a different file extensions on it to distinguish it from `ADCore.ibek.support.yaml`?

gilesknap replied:
Agreed on both.

gilesknap replied:
ADCore.install.yml ?

gilesknap commented:
I've changed the names of the ansible yml to <module>.install.yml and added schema for these files.

I'm ready to merge this non-breaking step now I think. I still have some more improvements for later, including the breaking change of removing the install.sh files. https://getupnote.com/share/notes/BbMdiHBsiGUm5BRiQfTWwm0iP9Q2/a30d29dd-ff80-4310-8b0a-a0162f989bb2

gilesknap commented:
Its Friday night and this is now beautiful enough to merge. 

It's fully working with ioc-adsimdetector and also working with about half of the modules in ioc-dlsvmevac. This includes ability to source the *.install.yml from ibek-support-dls. It also includes the ability to apply changes specific to the target architecture e.g. iocStats applies a git patch for RTEMS only.

From in the devcontainer you now have:-

- ansible.sh iocStats (or any support module)
- ansible.sh ioc
- ansible.sh all

Its all done via a single, elegant playbook: https://github.com/epics-containers/ibek-support/blob/119aa365f3c275d880c2f70c720004b0047d04cc/_ansible/playbook.yml#L1-L17
